HV-582, HV-960: Adding ability to pass class loader for loading user classes; Adding OSGi test

gunnarmorling opened the pull request:
This adds `HibernateValidatorConfiguration#userClassLoader()` for passing in a classloader to be used for loading all user-provided classes and resources (XML descriptors, classes named in XML descriptors, ValidationMessages). This is used in the new OSGi-based test for making classes/resources in the test bundle available to HV.

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorFactoryImpl.java:
```diff
@@ -125,6 +126,8 @@
 	 */
 	private final List<ValidatedValueUnwrapper<?>> validatedValueHandlers;
 
+	private final ClassLoader userClassLoader;
```
Should we null this in `close()` to avoid potential class loader leaks? References to the VF should be released after calling `close()`, but to be on the safe side it may be a good idea to set this to null here. It can't be final then any longer though which may be a problem.

gunnarmorling replied:
Same with some other references which hold on to the given class loader, e.g. `XmlMetaDataProvider`.

hferentschik replied:
> Should we null this in close() to avoid potential class loader leaks?

sounds reasonable 

>  It can't be final then any longer though which may be a problem.

What's your concern?

hferentschik replied:
Btw, do you think that `userClassLoader` is the most fitting name? Somehow I always think of 'end user' in this case. What's about `providedClassLoader`. It's maybe a bit more neutral.

gunnarmorling replied:
> What's your concern?

I'm wondering whether the field would have to be volatile then so it's safely published to other threads. I guess though @Sanne's "theorem of safe publication during construction phase" applies here as well.

gunnarmorling replied:
> Somehow I always think of 'end user' in this case.

Our users are application developers :)

> What's about providedClassLoader.

No big fan, isn't everything provided what is passed to the config? Maybe `externalClassLoader` or `userResourceClassLoader`? I think I still like `userClassLoader` best.

hferentschik replied:
> I'm wondering whether the field would have to be volatile then so it's safely published to other threads. I guess though @Sanne's "theorem of safe publication during construction phase" applies here as well.

I think during constructions it is safe. That was my understanding even before.

hferentschik replied:
> isn't everything provided what is passed to the config

Right, so it is making clear that this is not a class loader which we have access to per default, but it is "provided" to us during configuration.

> Maybe externalClassLoader

Could work. Better than userClassLoader imo

> userResourceClassLoader

-1

I think providedClassLoader is best :-)

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/internal/xml/ConstrainedExecutableBuilder.java:
```diff
@@ -55,7 +55,8 @@ private ConstrainedExecutableBuilder() {
 																			  String defaultPackage,
 																			  ParameterNameProvider parameterNameProvider,
 																			  ConstraintHelper constraintHelper,
-																			  AnnotationProcessingOptionsImpl annotationProcessingOptions) {
```
The static nature of the XML-related classes required to change many places to pass on the class loader. It'd be a bit easier if these were instance methods...

gunnarmorling commented on osgi/karaf-features/pom.xml:
```diff
@@ -0,0 +1,84 @@
+    <description>Hibernate Validator features for Apache Karaf</description>
+
+    <properties>
+        <maven.deploy.skip>true</maven.deploy.skip>
```
Should we actually deploy the feature file? I'd say yes. Any thoughts?

hferentschik replied:
I guess we could. In this case it is simple enough. I am reluctant on Searhc, since the whole osgi-fying of the Lucene parts seems a bit hacky. We don't have this here, so I guess we could consider this also as a new feature to provide a Karaf feature.xml

gunnarmorling replied:
Right, that's my thinking. There is a HV feature as part of Karaf's enterprise feature repo, but it'd still be useful to have our own, be it only to have one which always is up-to-date (the Karaf one is on HV 5.0.3 atm.).

hferentschik replied:
+1

cloudbees-pull-request-builder commented:
[HV-5-MASTER #629](https://hibernate-validator.ci.cloudbees.com/job/HV-5-MASTER/629/) SUCCESS
This pull request looks good

cloudbees-pull-request-builder commented:
[HV-5-MASTER #632](https://hibernate-validator.ci.cloudbees.com/job/HV-5-MASTER/632/) UNSTABLE
Looks like there's a problem with this pull request

cloudbees-pull-request-builder commented:
[HV-5-MASTER #634](https://hibernate-validator.ci.cloudbees.com/job/HV-5-MASTER/634/) SUCCESS
This pull request looks good

hferentschik commented:
What's about aligning a bit with Search and move this osgi modules into `integration`. Of course this means we need to create a new submodule there as well, eg `wildfly`. I think I like this grouping better and as said, it aligns with Search.

hferentschik commented:
I just got this: 

    -------------------------------------------------------
     T E S T S
    -------------------------------------------------------
    Running org.hibernate.validator.osgi.integrationtest.OsgiIntegrationTest
    Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 186.295 sec <<< FAILURE! - in org.hibernate.    validator.osgi.integrationtest.OsgiIntegrationTest
    org.hibernate.validator.osgi.integrationtest.OsgiIntegrationTest  Time elapsed: 185.878 sec  <<< ERROR!
    java.lang.RuntimeException: Cannot get the remote bundle context
    	at org.ops4j.pax.exam.rbc.client.intern.RemoteBundleContextClientImpl.getRemoteBundleContext(    RemoteBundleContextClientImpl.java:255)
    	at org.ops4j.pax.exam.rbc.client.intern.RemoteBundleContextClientImpl.waitForState(    RemoteBundleContextClientImpl.java:211)
    	at org.ops4j.pax.exam.karaf.container.internal.KarafTestContainer.waitForState(KarafTestContainer.    java:570)
    	at org.ops4j.pax.exam.karaf.container.internal.KarafTestContainer.startKaraf(KarafTestContainer.    java:230)
    	at org.ops4j.pax.exam.karaf.container.internal.KarafTestContainer.start(KarafTestContainer.java:180)
    	at org.ops4j.pax.exam.spi.reactors.EagerSingleStagedReactor.setUp(EagerSingleStagedReactor.java:86)
    	at org.ops4j.pax.exam.spi.reactors.EagerSingleStagedReactor.beforeClass(EagerSingleStagedReactor.    java:136)
    	at org.ops4j.pax.exam.spi.reactors.ReactorManager.beforeClass(ReactorManager.java:447)
    	at org.ops4j.pax.exam.junit.impl.ProbeRunner.run(ProbeRunner.java:97)
    	at org.ops4j.pax.exam.junit.PaxExam.run(PaxExam.java:93)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
    	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
    	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
    	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
    Caused by: java.rmi.NotBoundException: fce921c3-43a9-434e-911c-890bc57a39d1
    	at sun.rmi.registry.RegistryImpl.lookup(RegistryImpl.java:166)
    	at sun.rmi.registry.RegistryImpl_Skel.dispatch(Unknown Source)
    	at sun.rmi.server.UnicastServerRef.oldDispatch(UnicastServerRef.java:410)
    	at sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:268)
    	at sun.rmi.transport.Transport$1.run(Transport.java:178)
    	at sun.rmi.transport.Transport$1.run(Transport.java:175)
    	at java.security.AccessController.doPrivileged(Native Method)
    	at sun.rmi.transport.Transport.serviceCall(Transport.java:174)
    	at sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:557)
    	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:812)
    	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:671)
    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    	at java.lang.Thread.run(Thread.java:745)
    	at sun.rmi.transport.StreamRemoteCall.exceptionReceivedFromServer(StreamRemoteCall.java:276)
    	at sun.rmi.transport.StreamRemoteCall.executeCall(StreamRemoteCall.java:253)
    	at sun.rmi.server.UnicastRef.invoke(UnicastRef.java:379)
    	at sun.rmi.registry.RegistryImpl_Stub.lookup(Unknown Source)
    	at org.ops4j.pax.exam.rbc.client.intern.RemoteBundleContextClientImpl.getRemoteBundleContext(    RemoteBundleContextClientImpl.java:242)
    	at org.ops4j.pax.exam.rbc.client.intern.RemoteBundleContextClientImpl.waitForState(    RemoteBundleContextClientImpl.java:211)
    	at org.ops4j.pax.exam.karaf.container.internal.KarafTestContainer.waitForState(KarafTestContainer.    java:570)
    	at org.ops4j.pax.exam.karaf.container.internal.KarafTestContainer.startKaraf(KarafTestContainer.    java:230)
    	at org.ops4j.pax.exam.karaf.container.internal.KarafTestContainer.start(KarafTestContainer.java:180)
    	at org.ops4j.pax.exam.spi.reactors.EagerSingleStagedReactor.setUp(EagerSingleStagedReactor.java:86)
    	at org.ops4j.pax.exam.spi.reactors.EagerSingleStagedReactor.beforeClass(EagerSingleStagedReactor.    java:136)
    	at org.ops4j.pax.exam.spi.reactors.ReactorManager.beforeClass(ReactorManager.java:447)
    	at org.ops4j.pax.exam.junit.impl.ProbeRunner.run(ProbeRunner.java:97)
    	at org.ops4j.pax.exam.junit.PaxExam.run(PaxExam.java:93)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
    	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
    	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
    	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
    
    org.hibernate.validator.osgi.integrationtest.OsgiIntegrationTest  Time elapsed: 186.294 sec  <<< ERROR!
    java.lang.RuntimeException: Container never came up
    	at org.ops4j.pax.exam.karaf.container.internal.KarafTestContainer.stop(KarafTestContainer.java:540)
    	at org.ops4j.pax.exam.spi.reactors.EagerSingleStagedReactor.tearDown(EagerSingleStagedReactor.java:118)
    	at org.ops4j.pax.exam.spi.reactors.EagerSingleStagedReactor.afterClass(EagerSingleStagedReactor.    java:132)
    	at org.ops4j.pax.exam.spi.reactors.ReactorManager.afterClass(ReactorManager.java:431)
    	at org.ops4j.pax.exam.junit.impl.ProbeRunner.run(ProbeRunner.java:107)
    	at org.ops4j.pax.exam.junit.PaxExam.run(PaxExam.java:93)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
    
    
    Results :
    
    Tests in error: 
      OsgiIntegrationTest.org.hibernate.validator.osgi.integrationtest.OsgiIntegrationTest » Runtime
      JUnit4Provider.invoke:124->executeTestSet:153->execute:264 » Runtime Container...
    
    Tests run: 2, Failures: 0, Errors: 2, Skipped: 0
    
    [INFO] ------------------------------------------------------------------------
    [INFO] Reactor Summary:
    [INFO] 
    [INFO] Hibernate Validator Aggregator ..................... SUCCESS [  1.702 s]
    [INFO] Hibernate Validator Engine ......................... SUCCESS [01:35 min]
    [INFO] Hibernate Validator Engine JDK 1.8 Tests ........... SUCCESS [  3.815 s]
    [INFO] Hibernate Validator Portable Extension ............. SUCCESS [ 15.888 s]
    [INFO] Hibernate Validator TCK Runner ..................... SUCCESS [01:03 min]
    [INFO] Hibernate Validator Wildfly Integration Tests ...... SUCCESS [01:35 min]
    [INFO] Hibernate Validator Annotation Processor ........... SUCCESS [  8.743 s]
    [INFO] Hibernate Validator Performance Tests .............. SUCCESS [  1.181 s]
    [INFO] Hibernate Validator OSGi modules ................... SUCCESS [  0.080 s]
    [INFO] Hibernate Validator Karaf features ................. SUCCESS [  0.436 s]
    [INFO] Hibernate Validator OSGi integration test .......... FAILURE [04:48 min]
    [INFO] Hibernate Validator Manual ......................... SKIPPED
    [INFO] Hibernate Validator Distribution ................... SKIPPED
    [INFO] ------------------------------------------------------------------------
    [INFO] BUILD FAILURE
    [INFO] ------------------------------------------------------------------------
    [INFO] Total time: 09:34 min
    [INFO] Finished at: 2015-02-13T13:14:35+01:00
    [INFO] Final Memory: 72M/479M
    [INFO] ------------------------------------------------------------------------

CI seemed to have worked. I try a couple of times to see whether it is re-producable

hferentschik commented on engine/pom.xml:
```diff
@@ -229,10 +229,10 @@
                             javax.validation.*;version="[1.0.0,2.0.0)",
                             javax.xml.*;version="0",
-                            javax.el.*;version="[2.0.0,4.0.0)",
+                            javax.el.*;version="[2.0.0,4.0.0)";resolution:=optional,
```
nice

hferentschik commented on engine/pom.xml:
```diff
@@ -229,10 +229,10 @@
                             org.xml.sax.*;version="0",
                             org.jboss.logging.*;version="[3.1.0,4.0.0)",
-                            com.fasterxml.classmate.*;version="0.8.0",
+                            com.fasterxml.classmate.*;version="[${classmate.version},2.0.0)",
```
is it not safer to request a specific version?

gunnarmorling replied:
For a user intervals are nicer as it allows to work several clients with one and the same version of this dependency. Of course it depends on whether they evolve the API compatibly or not. I don't have a strong preference, only we should do it for other deps in the same way I'd say (see logging).

hferentschik replied:
Ok. I am no expert on this. Not sure what the recommended way would be. Happy to use ranges if you think it is nicer for the user. Makes sense (until it breaks).

Sanne replied:
Isn't the whole point of OSGi that - being modular - different modules can depend on different versions?

I always wished we could somehow make use of ranges in Maven dependency declarations (which are not really meant for modular environments) because of the reasons you state, still we don't as it's essentially impossible to test for it.

If any in the OSGi world I'd expect it to have the same drawbacks than in Maven, but less so a benefit.

gunnarmorling replied:
> Isn't the whole point of OSGi that - being modular - different modules can depend on different versions?

Sure, you can have different versions of the same module if needed. But of course its still nicer to only work with one version of possible - less classes loaded, less RAM used etc. Also using different versions ends once passing types via method signatures happens (which is not the case here for ClassMate, though). So using ranges generally is preferrable IMO in order to reduce the number of different versions of a lib in use. As said it depends on the compatability of the libs of course. OSGi puts strong emphasize on the semantics of versions, so ranges should work from this perspective. It's a different story of course how lib authors stick to it in reality.

hferentschik commented on engine/src/main/java/org/hibernate/validator/internal/xml/ClassLoadingHelper.java:
```diff
@@ -49,11 +48,7 @@
 	private ClassLoadingHelper() {
```
Wouldn't be nicer to make `ClassLoadingHelper` non static and pass an instance of it to the appropriate classes. Inside it can wrap the provided classloader and do the right thing.

gunnarmorling replied:
Yes, I guess so. I never was a fan of all these static methods in the XML package to begin with, but I didn't want to alter your design :) I can change it.

hferentschik commented:
FYI, three times in a row the same test failure. I am working on this branch rebased onto current master (in case that matters)

hferentschik commented on osgi/pom.xml:
```diff
@@ -0,0 +1,28 @@
+  ~ See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+  -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
```
What's the benefit of this intermediate pom?

gunnarmorling replied:
It allows to build everything OSGi in one go.

hferentschik replied:
Hmm, seems pretty useless to me, but ok.

gunnarmorling commented:
> What's about aligning a bit with Search and move this osgi modules into integration

It depends on what "integration" is supposed to mean. Is it as in "integration testing" or "integration with other techs"? In case of the former, it seems not a good fit if we want to deploy the Karaf feature (making it an actual deliverable and no testing-only stuff). In case of the latter, shouldn't the CDI extension also be moved to "integration"?

gunnarmorling commented:
> FYI, three times in a row the same test failure. I am working on this branch rebased onto current master (in case that matters)

Hum, that's strange. I'll rebase as well and try myself.

gunnarmorling commented:
Rebased locally, still passes with the latest master. Maybe you can find out something via debugging or in the log files?

hferentschik commented:
> It depends on what "integration" is supposed to mean. Is it as in "integration testing" or "integration with other techs"?

Testing - wildfly, osgi. That we publish the feature file is more of a side thing. 

> In case of the former, it seems not a good fit if we want to deploy the Karaf feature 

I don't think it matters

gunnarmorling commented:
> Testing - wildfly, osgi. That we publish the feature file is more of a side thing. 

I don't know, I find it awkward to have one of our deliverable artifacts within a testing module. If we publish it, the features file is no testing artifact any longer, it's really a "main" artifact.

hferentschik commented:
> I don't know, I find it awkward to have one of our deliverable artifacts within a testing module. 

I have no problem with it and I really would like to get some more structure into the directory layout. Also 'integration' does not directly imply tests only.

gunnarmorling commented:
> Also 'integration' does not directly imply tests only. 

Before you said something different: "Testing - wildfly, osgi.". Again, if it's meant more generally than testing, why not have the CDI extension in there as well?

> I really would like to get some more structure into the directory layout

Tbh. I don't see what the problem is. It is nicely structured as is, and I don't think "more structure" will help if it mixes differences concerns. What issue are we trying to solve here actually?

hferentschik commented:
> Before you said something different: "Testing - wildfly, osgi.". Again, if it's meant more generally than testing, why not have the CDI extension in there as well?

The directory is called 'integration' not 'testing', so my point was that it is a bit open for interpretation. I think I'd rather move cdi as well, than adding osgi on the top level, but that is not a must for me.

> What issue are we trying to solve here actually

It's to a great degree a matter of taste.

gunnarmorling commented:
Rebased and pushed several more commits:

* Renamed "userClassLoader" to "externalClassLoader"
* Enabled deployment of Karaf features.xml
* Simplified signatures of XML builder classes by passing in dependencies through constructor
* Doc update
* Some more minor clean-up

> so my point was that it is a bit open for interpretation.

Which IMO is a bit bad for such structure :) It should  give a clear indication about its purpose and where certain stuff should be located.

I've left the directory structure as is, as that's what is most sensible to me. If you feel strongly about it, feel free to change, but then let's move the PE to "integration" as well for the sake of consistency.

cloudbees-pull-request-builder commented:
[HV-5-MASTER #646](https://hibernate-validator.ci.cloudbees.com/job/HV-5-MASTER/646/) SUCCESS
This pull request looks good

hferentschik commented:
Thanks for the rework. The class loader stuff looks good now. However the OSGi test still fails:

    Running org.hibernate.validator.osgi.integrationtest.OsgiIntegrationTest
    Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 183.124 sec <<< FAILURE! - in org.hibernate.    validator.osgi.integrationtest.OsgiIntegrationTest
    org.hibernate.validator.osgi.integrationtest.OsgiIntegrationTest  Time elapsed: 182.833 sec  <<< ERROR!
    java.lang.RuntimeException: Cannot get the remote bundle context
    	at org.ops4j.pax.exam.rbc.client.intern.RemoteBundleContextClientImpl.getRemoteBundleContext(    RemoteBundleContextClientImpl.java:255)
    	at org.ops4j.pax.exam.rbc.client.intern.RemoteBundleContextClientImpl.waitForState(    RemoteBundleContextClientImpl.java:211)
    	at org.ops4j.pax.exam.karaf.container.internal.KarafTestContainer.waitForState(KarafTestContainer.    java:570)
    	at org.ops4j.pax.exam.karaf.container.internal.KarafTestContainer.startKaraf(KarafTestContainer.    java:230)
    	at org.ops4j.pax.exam.karaf.container.internal.KarafTestContainer.start(KarafTestContainer.java:180)
    	at org.ops4j.pax.exam.spi.reactors.EagerSingleStagedReactor.setUp(EagerSingleStagedReactor.java:86)
    	at org.ops4j.pax.exam.spi.reactors.EagerSingleStagedReactor.beforeClass(EagerSingleStagedReactor.    java:136)
    	at org.ops4j.pax.exam.spi.reactors.ReactorManager.beforeClass(ReactorManager.java:447)
    	at org.ops4j.pax.exam.junit.impl.ProbeRunner.run(ProbeRunner.java:97)
    	at org.ops4j.pax.exam.junit.PaxExam.run(PaxExam.java:93)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
    	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
    	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
    	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
    Caused by: java.rmi.NotBoundException: d60e3159-abe2-4497-9443-69eec7115c12
    	at sun.rmi.registry.RegistryImpl.lookup(RegistryImpl.java:166)
    	at sun.rmi.registry.RegistryImpl_Skel.dispatch(Unknown Source)
    	at sun.rmi.server.UnicastServerRef.oldDispatch(UnicastServerRef.java:410)
    	at sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:268)
    	at sun.rmi.transport.Transport$1.run(Transport.java:178)
    	at sun.rmi.transport.Transport$1.run(Transport.java:175)
    	at java.security.AccessController.doPrivileged(Native Method)
    	at sun.rmi.transport.Transport.serviceCall(Transport.java:174)
    	at sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:557)
    	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:812)
    	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:671)
    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    	at java.lang.Thread.run(Thread.java:745)
    	at sun.rmi.transport.StreamRemoteCall.exceptionReceivedFromServer(StreamRemoteCall.java:276)
    	at sun.rmi.transport.StreamRemoteCall.executeCall(StreamRemoteCall.java:253)
    	at sun.rmi.server.UnicastRef.invoke(UnicastRef.java:379)
    	at sun.rmi.registry.RegistryImpl_Stub.lookup(Unknown Source)
    	at org.ops4j.pax.exam.rbc.client.intern.RemoteBundleContextClientImpl.getRemoteBundleContext(    RemoteBundleContextClientImpl.java:242)
    	at org.ops4j.pax.exam.rbc.client.intern.RemoteBundleContextClientImpl.waitForState(    RemoteBundleContextClientImpl.java:211)
    	at org.ops4j.pax.exam.karaf.container.internal.KarafTestContainer.waitForState(KarafTestContainer.    java:570)
    	at org.ops4j.pax.exam.karaf.container.internal.KarafTestContainer.startKaraf(KarafTestContainer.    java:230)
    	at org.ops4j.pax.exam.karaf.container.internal.KarafTestContainer.start(KarafTestContainer.java:180)
    	at org.ops4j.pax.exam.spi.reactors.EagerSingleStagedReactor.setUp(EagerSingleStagedReactor.java:86)
    	at org.ops4j.pax.exam.spi.reactors.EagerSingleStagedReactor.beforeClass(EagerSingleStagedReactor.    java:136)
    	at org.ops4j.pax.exam.spi.reactors.ReactorManager.beforeClass(ReactorManager.java:447)
    	at org.ops4j.pax.exam.junit.impl.ProbeRunner.run(ProbeRunner.java:97)
    	at org.ops4j.pax.exam.junit.PaxExam.run(PaxExam.java:93)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
    	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
    	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
    	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
    
    org.hibernate.validator.osgi.integrationtest.OsgiIntegrationTest  Time elapsed: 183.124 sec  <<< ERROR!
    java.lang.RuntimeException: Container never came up
    	at org.ops4j.pax.exam.karaf.container.internal.KarafTestContainer.stop(KarafTestContainer.java:540)
    	at org.ops4j.pax.exam.spi.reactors.EagerSingleStagedReactor.tearDown(EagerSingleStagedReactor.java:118)
    	at org.ops4j.pax.exam.spi.reactors.EagerSingleStagedReactor.afterClass(EagerSingleStagedReactor.    java:132)
    	at org.ops4j.pax.exam.spi.reactors.ReactorManager.afterClass(ReactorManager.java:431)
    	at org.ops4j.pax.exam.junit.impl.ProbeRunner.run(ProbeRunner.java:107)
    	at org.ops4j.pax.exam.junit.PaxExam.run(PaxExam.java:93)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
    	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
    
    
    Results :
    
    Tests in error: 
      OsgiIntegrationTest.org.hibernate.validator.osgi.integrationtest.OsgiIntegrationTest » Runtime
      JUnit4Provider.invoke:124->executeTestSet:153->execute:264 » Runtime Container...
    
    Tests run: 2, Failures: 0, Errors: 2, Skipped: 0

For what it's worth, I ran the Search OSGi test and that works fine. I don't know what the difference is atm. Have you by any chance tried running the test on OS X?

To move this a bit along, I have split this pull request into two sets of commits - one for HV-582 and one for HV-960 (quite some fiddling). I am going to apply the changes for HV-582 and create a new pull request with the other changes.

gunnarmorling commented:
> To move this a bit along, I have split this pull request into two sets of commits - one for HV-582 and one for HV-960 (quite some fiddling). I am going to apply the changes for HV-582 and create a new pull request with the other changes. 

Mh, tbh. I'm a bit unhappy about this handling.

You squashed commits which I had separated on purpose. It makes sense for the review feedback commits (I would have squashed those myself once we had agreed that no more changes are required). But there is one large commit now for the XML classes which mixes the classloader usage and a change independent from that (passing the global state through constructors). These should have stayed separate.

Also it'd be nice to keep the original author of the change. You can do so via git commit --author="Name <email>".
